File: java/215.kth-largest-element-in-an-array.java
/*
 * @lc app=leetcode id=215 lang=java
 *
 * [215] Kth Largest Element in an Array
 */
class Solution {

    public int findKthLargest(int[] nums, int k) {

        k = nums.length - k;
        int lo = 0;
        int hi = nums.length - 1;
        while (lo < hi) {
            final int j = partition(nums, lo, hi);
            if(j < k) {
                lo = j + 1;
            } else if (j > k) {
                hi = j - 1;
            } else {
                break;
            }
        }
        return nums[k];
    }

    private int partition(int[] a, int lo, int hi) {
        int i = lo;
        int j = hi + 1;
        while(true) {
            while(i < hi && a[++i] < a[lo]);
            while(j > lo && a[lo] < a[--j]);
            if(i >= j) break;
            swap(a, i, j);
        }
        swap(a, lo, j);
        return j;
    }

    private void swap(int[] a, int i, int j) {
        final int tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }
}

File: go/215.kth-largest-element-in-an-array.go
/*
 * @lc app=leetcode id=215 lang=golang
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
func findKthLargest(nums []int, k int) int {
	k = len(nums) - k
	lo, hi := 0, len(nums)-1
	for lo < hi {
		sep := split(nums, lo, hi)
		if sep < k {
			lo = sep + 1
		} else if sep > k {
			hi = sep - 1
		} else {
			break
		}
	}
	return nums[k]
}

func split(nums []int, lo int, hi int) int {
	i, j := lo, hi+1
	for {
		for i < hi {
			i++
			if !(nums[i] < nums[lo]) {
				break
			}
		}
		for lo < j {
			j--
			if !(nums[lo] < nums[j]) {
				break
			}
		}
		if i >= j {
			break
		}
		nums[i], nums[j] = nums[j], nums[i]
	}
	nums[lo], nums[j] = nums[j], nums[lo]
	return j
}

// @lc code=end

